### DISCUSSION TITLE
Try new Improved epub-reader beta

### DISCUSSION THREAD

**Discussion by mienaiyami** [mienaiyami/yomikiru#384] [Announcements]
# Try and test new improved epub-reader

Executable files at https://github.com/mienaiyami/yomikiru-improved-epubreader-files/releases/
Code at https://github.com/mienaiyami/yomikiru/tree/better-epub-reader

> This is only a temporary version meant for testing to find major issues before release.

---

## Note

If you use Yomikiru's EPUB reader, you can improve it by providing feedback while I'm working on a more enhanced version of it. 
What can you do? You can mention bugs in the executable above, you can also provide suggestion to improve existing features, mention anything that is missing.

**Do not create new issues for them, give description only below this discussion.**

## What changed?
In 1 or 2 weeks Yomikiru's epub-reader will be replaced with a new one. Though it will still have existing features (most) but new are also added for better experience.
Frontend might not be that different from previous versions, but you can expect enhanced performance because a lot of code changes were done on fundamental level.
Change list:
- enhanced performance when scrolling, moving through chapters.
- enhanced formatting
- enhanced internal link following
- improved history/bookmark saving
- improved "find in page"
- more secure
- new features (not implemented yet): 
  - notes, highlight text and add note with color and additional text
  - better bookmark, bookmarks will be moved inside epub side-list allowing multiple bookmarks in same epub.

## Why?
There were many reasons that made me redo the epub-reader:
- previous version of epub-reader was hastly written and wasn't well thought.
- I also lacked enough experience with epub when I first wrote it.
- there were major performance issues with large epubs (1000+ chapters).
- hard to implement new features.
- and many more I forgot now.

**Comment by mienaiyami** [comment]
# Subscribe to this discussion to get future updates till final release.

Note:
For epub with chapters > 500, click on "Content" in side-list to hide it and increase performance drastically.

**Comment by mienaiyami** [comment]
# updated to v1.1-beta
- fixed scroll performance issue.

**Reply by zeoint** [reply]
Other than coming soon features, Everything in the epub reader is working smooth including the chapter transition. I encountered no bugs.

**Comment by zeoint** [comment]
Found a bug, opened at [beta issues](https://github.com/mienaiyami/yomikiru-improved-epubreader-files/issues/1).

**Reply by mienaiyami** [reply]
Thank you for reporting issues.
I haven't been able to work on Yomikiru for a long time because of some other works, but I will soon resume it.

**Reply by zeoint** [reply]
Cool.

**Comment by mienaiyami** [comment]
I may soon resume work on Yomikiru, and I want to discuss some upcoming changes that I've been considering for a while.

## Your bookmarks in Yomikiru may stop working in future updates.

To keep it short, I hit a roadblock while enhancing the epub reader; I couldn't integrate notes and internal bookmarks due to the bookmark and history file system. Since Yomikiru still relies on the original system designed for manga/comics, which was very basic, it's become challenging to incorporate new features with the addition of epubs. It's not certain, but there's a high chance that bookmarks may become invalid because trying to save them for next version is not that simple.
Furthermore, some users may have noticed planned updates to simplify the UI (https://github.com/mienaiyami/yomikiru/labels/gallery%20mode). The current user data structure is causing problems in that as well, so I have planned to redesign it.

Your history/continue-reading data will remain unaffected.

**Reply by omodia** [reply]
What if list and gallery view both are based in the same UI?

### 1. List and Gallery View
- Layout = Drop down menu containing View modes, Row size and Sorting Options.
- Ask for confirmation to change between view modes.
- Row Size = Ranging from 2 to 8 or 10. Cycle through range size on clicking row size.
- Perma show search bar on all three tabs.
- Show layout option only on location tab, if you don't plan to implement gallery view for bookmarks and continue reading tab.

![UI](https://github.com/mienaiyami/yomikiru/assets/116664905/a674d751-edea-49bb-af60-16044bb032cd)

**Reply by mienaiyami** [reply]
@omodia 
That is not the issue.
Actual problem lies in the way user data is stored, making it hard to change or add anything new.

**Reply by omodia** [reply]
I understood that from your comment.
Updated Above Image.
I was suggesting a UI layout I thought of that doesn't change much when shifting between view modes.

**Reply by omodia** [reply]
You could freeze unnecessary options when shifting between the view modes in this UI.
What do you think?
- List View
- 
![List View](https://github.com/mienaiyami/yomikiru/assets/116664905/8f8b4520-0941-402c-8320-af931df68e8a)

---

- Gallery View
- 
![Gallery View](https://github.com/mienaiyami/yomikiru/assets/116664905/861e4f9e-ab15-4563-bf78-4d69f1f2799f)

---

- Cover View = Gallery View, but names hidden
- 
![Cover View](https://github.com/mienaiyami/yomikiru/assets/116664905/68f3ea4b-f18d-4470-a74f-acb63d0b05f9)

**Reply by mienaiyami** [reply]
There are still many things to do before I can even start implementing this UI. And with my other workload increased, it will take few months before I can even start this.

**Comment by zeoint** [comment]
Cool Home screen UI.
By the way, the comic and epub reader UI, imo, is as perfect as it can be.
